src/main/java/hintahaukka/database/Database.java

```diff
@@ -51,23 +23,26 @@ public Connection getConnection() throws URISyntaxException, SQLException {
         return DriverManager.getConnection(dbUrl, username, password);
     }
     
-    public void initializeDatabaseIfUninitialized() throws URISyntaxException, SQLException {
-        if (countTables() != CREATE_TABLE_STATEMENTS.length) {
-            executeStatements(CREATE_TABLE_STATEMENTS);
+    public void initializeDatabaseIfUninitialized(String schemaName) throws URISyntaxException, SQLException {
+        String[] statements = createTableStatements(schemaName);
+        if (countTables(schemaName) != statements.length) {
+            executeStatements(new String[]{"CREATE SCHEMA " + schemaName});
+            executeStatements(statements);
         }
     }
     
-    public void clearDatabase() throws URISyntaxException, SQLException {
-        if (countTables() == CREATE_TABLE_STATEMENTS.length) {
-            executeStatements(DROP_TABLE_STATEMENTS);
+    public void clearDatabase(String schemaName) throws URISyntaxException, SQLException {
+        String[] statements = createTableStatements(schemaName);
+        if (countTables(schemaName) == statements.length) {
+            executeStatements(new String[]{"DROP SCHEMA " + schemaName + " CASCADE"});
         }        
     }    
     
     
-    int countTables() throws URISyntaxException, SQLException {
+    int countTables(String schemaName) throws URISyntaxException, SQLException {
         Connection connection = this.getConnection();
         PreparedStatement statement = connection.prepareStatement(
-            "SELECT count(*) FROM pg_stat_user_tables"
+            "SELECT count(*) FROM pg_stat_user_tables WHERE schemaname = '" + schemaName + "'"
```

Comment:
Onko tässä (ja muissa kohdissa tän jälkeen) SQL-injektion vaara?

Reply:
Ei ole :)

SQL-injektio kävi kyllä mielessä tuota tehdessä. Jos käyttäjällä olisi mahdollista päästä syöttämään dataa tuohon muuttujaan niin olisin käyttänyt "? ja setString"-tapaa, mutta käyttäjä voi toiminnallaan vain valita tuleeko tuohon "public" vai "test".

Routeissa joko syötetään tuo käyttäjän puolesta, paitsi reset-routissa käyttäjä syöttää, mutta tällöin käyttäjän syöte tarkistetaan että se on "public" tai "test".

Mutta javadoc-kommenttiin minun olisi pitänyt laittaa muistutuksena muille ja toki myös itselle jos pääsee itseltä unohtumaan, että tuohon ei saa koskaan päästää käyttäjän vapaata dataa. Lisään kommentin kun teen paremmat javadocit jossain vaiheessa.

Reply:
Okei, ajattelinkin että olit varmaan miettinyt tän kun kerran tuolla oli muuten varottu injektioita. Hyvä homma :+1: